src/Storages/StorageMergeTree.cpp

```diff
@@ -1140,12 +1139,38 @@ Pipe StorageMergeTree::alterPartition(
     return {};
 }
 
+
+ActionLock StorageMergeTree::stopMergesAndWait()
+{
+    /// Asks to complete merges and does not allow them to start.
+    /// This protects against "revival" of data for a removed partition after completion of merge.
+    auto merge_blocker = merger_mutator.merges_blocker.cancel();
+
+    {
```

Comment:
This is what changed. We have not only to ask merges for cancel. But to actually wait that they are finished.